plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/embeddings/TypeEmbedding.kt

```diff
@@ -231,47 +236,82 @@ data class ClassTypeEmbedding(val className: ScopedKotlinName, val isInterface:
+    val sharedPredicate: Predicate
+        get() = _sharedPredicate ?: error("Predicate of $className has not been initialised yet.")
+    val uniquePredicate: Predicate
+        get() = _uniquePredicate ?: error("Unique Predicate of $className has not been initialised yet.")
 
     fun initFields(newFields: Map<SimpleKotlinName, FieldEmbedding>) {
         check(_fields == null) { "Fields of $className are already initialised." }
         _fields = newFields
-        _predicate = initPredicate()
+        _sharedPredicate = initSharedPredicate()
+        _uniquePredicate = initUniquePredicate()
     }
 
-    private fun initPredicate(): Predicate {
-        val subjectEmbedding = PlaceholderVariableEmbedding(ClassPredicateSubjectName, this)
-        val accessFields = fields.values
-            .mapNotNull { it.accessInvariantsForPredicate()?.fillHole(subjectEmbedding) }
+    private val subjectEmbedding = PlaceholderVariableEmbedding(ClassPredicateSubjectName, this)
+    private fun initSharedPredicate(): Predicate {
+        val filteredFields = fields.values
+            .filterIsInstance<UserFieldEmbedding>()
+            .filter { it.accessPolicy == AccessPolicy.ALWAYS_READABLE }
+
+        val accessFields = filteredFields.map { FieldAccessTypeInvariantEmbedding(it, PermExp.WildcardPerm()).fillHole(subjectEmbedding) }
 
-        // For the moment ALWAYS_WRITEABLE fields with some class type do not exist.
-        // Whether to include them here will need to be considered in the future.
         // We need to pass in the field access since the predicates are simply the ones for the type.
-        val accessFieldPredicates = fields.values
-            .filter { it.accessPolicy == AccessPolicy.ALWAYS_READABLE }
-            .mapNotNull { it.type.predicateAccessInvariant()?.fillHole(PrimitiveFieldAccess(subjectEmbedding, it)) }
+        val accessFieldPredicates = filteredFields
+            .mapNotNull { it.type.sharedPredicateAccessInvariant()?.fillHole(PrimitiveFieldAccess(subjectEmbedding, it)) }
 
-        val fieldsProvenInvariants = fields.values
-            .filter { it.accessPolicy == AccessPolicy.ALWAYS_READABLE }
+        val fieldsProvenInvariants = filteredFields
             .flatMap { it.type.provenInvariants().fillHoles(PrimitiveFieldAccess(subjectEmbedding, it)) }
 
-        val accessSuperTypesPredicates = superTypes
-            .filterIsInstance<ClassTypeEmbedding>()
+        val accessSuperTypesPredicates = classSuperTypes
             .map { PredicateAccessTypeInvariantEmbedding(it.name, PermExp.WildcardPerm()).fillHole(subjectEmbedding) }
 
         val body = (accessFields + accessFieldPredicates + accessSuperTypesPredicates + fieldsProvenInvariants).toConjunction()
         return Predicate(name, listOf(subjectEmbedding.toLocalVarDecl()), body.pureToViper(toBuiltin = true))
     }
 
+    private fun initUniquePredicate(): Predicate {
+        val filteredFields = fields.values.filterIsInstance<UserFieldEmbedding>()
```

Comment:
Why is it not necessary to exclude always-readable fields here?

Reply:
Let's consider this example:
```kt
class A(@unique val x: T)
```
Here we need to include `x` because we want to access its unique predicate
```
UniqueA(this: Ref) { acc(this.x, wildcard) && UniqueT(this.x) }
```

Reply:
yes, probably using a predicate builder is a good idea. I'll try to implement it.